File: src/main/java/com/extractor/model/Component.java
package com.extractor.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;
import java.util.ArrayList;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Represents a component (class/interface/enum) in the dependency graph.
 * This matches the Sofka schema exactly - simple lists for calls_out/calls_in,
 * no dependencies field.
 */
public class Component {

    @JsonProperty("id")
    private String id; // Fully qualified class name

    @JsonProperty("files")
    private List<String> files;

    @JsonProperty("loc")
    private int loc; // Lines of code

    @JsonProperty("tables_used")
    private List<String> tablesUsed;

    @JsonProperty("sensitive_data")
    private boolean sensitiveData;

    @JsonProperty("domain")
    private String domain;

    @JsonProperty("layer")
    private String layer;

    @JsonProperty("calls_out")
    private List<String> callsOut;

    @JsonProperty("calls_in")
    private List<String> callsIn;

    @JsonProperty("ejb_type")
    private String ejbType;

    @JsonProperty("uses_jndi")
    private boolean usesJNDI;

    @JsonProperty("annotations")
    private List<String> annotations;

    @JsonProperty("is_interface")
    private boolean isInterface;

    @JsonProperty("secrets_references")
    private List<String> secretsReferences;

    @JsonProperty("external_dependencies")
    private List<String> externalDependencies;

    @JsonProperty("package_dependencies")
    private List<PackageGroup> packageDependencies;

    @JsonProperty("code_issues")
    private List<CodeIssue> codeIssues;

    @JsonProperty("extends")
    private String extendsClass;

    @JsonProperty("implements")
    private List<String> implementsInterfaces;

    @JsonProperty("messaging_type")
    private String messagingType;

    @JsonProperty("messaging_role")
    private String messagingRole;

    @JsonProperty("web_type")
    private String webType;

    @JsonProperty("web_role")
    private String webRole;

    @JsonProperty("cbo")
    private Integer cbo; // Coupling Between Objects

    @JsonProperty("lcom")
    private Double lcom; // Lack of Cohesion in Methods (LCOM-HS: 0=high cohesion, 1=low cohesion)

    public Component() {
        this.files = new ArrayList<>();
        this.tablesUsed = new ArrayList<>();
        this.callsOut = new ArrayList<>();
        this.callsIn = new ArrayList<>();
        this.annotations = new ArrayList<>();
        this.secretsReferences = new ArrayList<>();
        this.externalDependencies = new ArrayList<>();
        this.packageDependencies = new ArrayList<>();
        this.implementsInterfaces = new ArrayList<>();
        this.codeIssues = new ArrayList<>();
        this.sensitiveData = false;
        this.usesJNDI = false;
        this.isInterface = false;
    }

    public Component(String id) {
        this();
        this.id = id;
    }

    // Getters and setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void addFile(String file) {
        if (!this.files.contains(file)) {
            this.files.add(file);
        }
    }

    public List<String> getFiles() {
        return files;
    }

    public List<CodeIssue> getCodeIssues() {
        return codeIssues;
    }

    public void setCodeIssues(List<CodeIssue> codeIssues) {
        this.codeIssues = codeIssues;
    }

    public void addCodeIssue(CodeIssue issue) {
        if (this.codeIssues == null) {
            this.codeIssues = new ArrayList<>();
        }
        this.codeIssues.add(issue);
    }

    public int getLoc() {
        return loc;
    }

    public void setLoc(int loc) {
        this.loc = loc;
    }

    public List<String> getTablesUsed() {
        return tablesUsed;
    }

    public void setTablesUsed(List<String> tablesUsed) {
        this.tablesUsed = tablesUsed;
    }

    public void addTableUsed(String tableName) {
        if (!this.tablesUsed.contains(tableName)) {
            this.tablesUsed.add(tableName);
        }
    }

    public boolean isSensitiveData() {
        return sensitiveData;
    }

    public void setSensitiveData(boolean sensitiveData) {
        this.sensitiveData = sensitiveData;
    }

    public String getDomain() {
        return domain;
    }

    public void setDomain(String domain) {
        this.domain = domain;
    }

    public String getLayer() {
        return layer;
    }

    public void setLayer(String layer) {
        this.layer = layer;
    }

    public List<String> getCallsOut() {
        return callsOut;
    }

    public void setCallsOut(List<String> callsOut) {
        this.callsOut = callsOut;
    }

    public void addCallOut(String to) {
        if (!callsOut.contains(to)) {
            callsOut.add(to);
        }
    }

    public List<String> getCallsIn() {
        return callsIn;
    }

    public void setCallsIn(List<String> callsIn) {
        this.callsIn = callsIn;
    }

    public void addCallIn(String from) {
        if (!callsIn.contains(from)) {
            callsIn.add(from);
        }
    }

    public String getEjbType() {
        return ejbType;
    }

    public void setEjbType(String ejbType) {
        this.ejbType = ejbType;
    }

    public boolean isUsesJNDI() {
        return usesJNDI;
    }

    public void setUsesJNDI(boolean usesJNDI) {
        this.usesJNDI = usesJNDI;
    }

    public List<String> getAnnotations() {
        return annotations;
    }

    public void setAnnotations(List<String> annotations) {
        this.annotations = annotations;
    }

    public void addAnnotation(String annotation) {
        if (!this.annotations.contains(annotation)) {
            this.annotations.add(annotation);
        }
    }

    public boolean isInterface() {
        return isInterface;
    }

    public void setInterface(boolean isInterface) {
        this.isInterface = isInterface;
    }

    public List<String> getSecretsReferences() {
        return secretsReferences;
    }

    public void setSecretsReferences(List<String> secretsReferences) {
        this.secretsReferences = secretsReferences;
    }

    public List<String> getExternalDependencies() {
        return externalDependencies;
    }

    public void setExternalDependencies(List<String> externalDependencies) {
        this.externalDependencies = externalDependencies;
    }

    public void addExternalDependency(String dependency) {
        if (!this.externalDependencies.contains(dependency)) {
            this.externalDependencies.add(dependency);
        }
    }

    public List<PackageGroup> getPackageDependencies() {
        return packageDependencies;
    }

    public void setPackageDependencies(List<PackageGroup> packageDependencies) {
        this.packageDependencies = packageDependencies;
    }

    public void addPackageDependency(PackageGroup packageGroup) {
        if (packageGroup != null && !this.packageDependencies.contains(packageGroup)) {
            this.packageDependencies.add(packageGroup);
        }
    }

    public String getExtendsClass() {
        return extendsClass;
    }

    public void setExtendsClass(String extendsClass) {
        this.extendsClass = extendsClass;
    }

    public List<String> getImplementsInterfaces() {
        return implementsInterfaces;
    }

    public void setImplementsInterfaces(List<String> implementsInterfaces) {
        this.implementsInterfaces = implementsInterfaces;
    }

    public void addImplementsInterface(String interfaceName) {
        if (!this.implementsInterfaces.contains(interfaceName)) {
            this.implementsInterfaces.add(interfaceName);
        }
    }

    public String getMessagingType() {
        return messagingType;
    }

    public void setMessagingType(String messagingType) {
        this.messagingType = messagingType;
    }

    public String getMessagingRole() {
        return messagingRole;
    }

    public void setMessagingRole(String messagingRole) {
        this.messagingRole = messagingRole;
    }

    public String getWebType() {
        return webType;
    }

    public void setWebType(String webType) {
        this.webType = webType;
    }

    public String getWebRole() {
        return webRole;
    }

    public void setWebRole(String webRole) {
        this.webRole = webRole;
    }

    public Integer getCbo() {
        return cbo;
    }

    public void setCbo(Integer cbo) {
        this.cbo = cbo;
    }

    public Double getLcom() {
        return lcom;
    }

    public void setLcom(Double lcom) {
        this.lcom = lcom;
    }

    /**
     * Normalize the component data by sorting and deduplicating collections.
     */
    public void normalize() {
        // Sort and deduplicate tables_used
        tablesUsed = tablesUsed.stream().distinct().sorted().collect(Collectors.toList());

        // Sort and deduplicate calls_out and calls_in
        callsOut = callsOut.stream().distinct().sorted().collect(Collectors.toList());
        callsIn = callsIn.stream().distinct().sorted().collect(Collectors.toList());

        // Sort and deduplicate annotations
        annotations = annotations.stream().distinct().sorted().collect(Collectors.toList());

        // Sort and deduplicate secrets references
        secretsReferences = secretsReferences.stream().distinct().sorted().collect(Collectors.toList());

        // Sort and deduplicate external dependencies
        externalDependencies = externalDependencies.stream().distinct().sorted().collect(Collectors.toList());

        // Sort and deduplicate implements interfaces
        implementsInterfaces = implementsInterfaces.stream().distinct().sorted().collect(Collectors.toList());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Component component = (Component) o;
        return Objects.equals(id, component.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Component{" +
                "id='" + id + '\'' +
                ", files=" + files +
                ", loc=" + loc +
                ", tablesUsed=" + tablesUsed +
                ", sensitiveData=" + sensitiveData +
                '}';
    }
}

File: src/main/dotnet/DotNetAnalyzer/Model/Component.cs
using System.Text.Json.Serialization;

namespace DotNetAnalyzer.Model;

public class Component
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("files")]
    public List<string> Files { get; set; }

    [JsonPropertyName("loc")]
    public int Loc { get; set; }

    [JsonPropertyName("tables_used")]
    public List<string> TablesUsed { get; set; }

    [JsonPropertyName("sensitive_data")]
    public bool SensitiveData { get; set; }

    [JsonPropertyName("domain")]
    public string? Domain { get; set; }

    [JsonPropertyName("layer")]
    public string? Layer { get; set; }

    [JsonPropertyName("calls_out")]
    public List<string> CallsOut { get; set; }

    [JsonPropertyName("calls_in")]
    public List<string> CallsIn { get; set; }

    [JsonPropertyName("ejb_type")]
    public string? EjbType { get; set; }

    [JsonPropertyName("uses_jndi")]
    public bool UsesJndi { get; set; }

    [JsonPropertyName("annotations")]
    public List<string> Annotations { get; set; }

    [JsonPropertyName("is_interface")]
    public bool IsInterface { get; set; }

    [JsonPropertyName("secrets_references")]
    public List<string> SecretsReferences { get; set; }

    [JsonPropertyName("external_dependencies")]
    public List<string> ExternalDependencies { get; set; }

    [JsonPropertyName("package_dependencies")]
    public List<PackageGroup> PackageDependencies { get; set; }

    [JsonPropertyName("code_issues")]
    public List<CodeIssue> CodeIssues { get; set; }

    [JsonPropertyName("extends")]
    public string? Extends { get; set; }

    [JsonPropertyName("implements")]
    public List<string> Implements { get; set; }

    [JsonPropertyName("messaging_type")]
    public string? MessagingType { get; set; }

    [JsonPropertyName("messaging_role")]
    public string? MessagingRole { get; set; }

    [JsonPropertyName("web_type")]
    public string? WebType { get; set; }

    [JsonPropertyName("web_role")]
    public string? WebRole { get; set; }

    [JsonPropertyName("cbo")]
    public int? Cbo { get; set; }

    [JsonPropertyName("lcom")]
    public double? Lcom { get; set; }

    public Component()
    {
        Files = new List<string>();
        TablesUsed = new List<string>();
        CallsOut = new List<string>();
        CallsIn = new List<string>();
        Annotations = new List<string>();
        SecretsReferences = new List<string>();
        ExternalDependencies = new List<string>();
        PackageDependencies = new List<PackageGroup>();
        CodeIssues = new List<CodeIssue>();
        Implements = new List<string>();
    }
}

public class PackageGroup
{
    [JsonPropertyName("group")]
    public string Group { get; set; } = string.Empty;

    [JsonPropertyName("artifact")]
    public string Artifact { get; set; } = string.Empty;

    [JsonPropertyName("version")]
    public string? Version { get; set; }
}

public class CodeIssue
{
    /// <summary>Rule identifier (C# API). Serialized as "pattern" for Java compatibility.</summary>
    [JsonPropertyName("pattern")]
    public string Rule { get; set; } = string.Empty;

    [JsonPropertyName("type")]
    public string Type { get; set; } = "BUG_PATTERN";

    [JsonPropertyName("severity")]
    public string Severity { get; set; } = string.Empty;

    [JsonPropertyName("message")]
    public string Message { get; set; } = string.Empty;

    [JsonPropertyName("line")]
    public int Line { get; set; }

    /// <summary>Human-readable location (file:line). Not serialized to JSON.</summary>
    [JsonIgnore]
    public string? Location { get; set; }
}
